### PR TITLE
Bump Docsy to 0.4.x

### PR DESCRIPTION
**Pull request by sftim**
This PR upgrades to a newer version of Docsy: 0.4.0 ([preview](https://deploy-preview-48722--kubernetes-io-main-staging.netlify.app/docs/))

Fixes https://github.com/kubernetes/website/issues/32905; also relevant to issue https://github.com/kubernetes/website/issues/44002

Importantly, the container image now installs dependencies using NPM, _not_ a Git submodule.
In the testing I've done, I haven't found any different look or behavior that still needs fixing.

/area web-development

This PR incorporates the commits from PR #48721

### CODE REVIEW

**Review by Arhell — COMMENTED** [review]
/lgtm

**Review by chalin — COMMENTED** [review]
With the changes suggested below, we can then successfully run both local and container build and serve.

File: Dockerfile [hunk 1]
```diff
@@ -1,47 +1,37 @@
...
 
-RUN apk add --no-cache \
-    runuser \
-    git \
-    openssh-client \
-    rsync \
-    npm && \
-    npm install -D autoprefixer postcss-cli
+WORKDIR /opt/npm
+RUN npm install -D -g autoprefixer postcss-cli google/docsy#semver:0.4.0
```
[the remark this answers is not in the document]
Comment (chalin) [reply Dockerfile:29]: I actually found a better way (IMHO), which copies the project's packages files rather that duplicating dependencies here. I'll share that tomorrow, I'm calling it a day.

**Comment by chalin** [comment]
Actually, here's the [proposal I made earlier](https://github.com/kubernetes/website/pull/48722#discussion_r1844827660) that copies the website's NPM package files for use in the container image; this avoids a global install (which isn't usually a good thing for non CLI packages), and  keeps things more DRY:

```diff
diff --git a/.dockerignore b/.dockerignore
index 1d085cacc9..be2f2c26c4 100644
--- a/.dockerignore
+++ b/.dockerignore
@@ -1 +1,3 @@
 **
+!package.json
+!package-lock.json
```
```diff
diff --git a/Dockerfile b/Dockerfile
index 9e07877af0..9cde91c750 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -25,8 +25,11 @@ RUN apt-get update && apt-get install -y \
 
 RUN rm -rf /var/cache/* # partial cleanup
 
-WORKDIR /opt/npm
-RUN npm install -g autoprefixer postcss-cli google/docsy#semver:0.4.0
+RUN mkdir -p /npm-dep
+COPY package.json package-lock.json /npm-dep
+
+WORKDIR /npm-dep
+RUN npm install
 
 RUN useradd -m --user-group -u 60000 -d /var/hugo hugo && \
     chown -R hugo: /var/hugo && \
```
```diff
diff --git a/Makefile b/Makefile
index e5716e1193..c9e52333da 100644
--- a/Makefile
+++ b/Makefile
@@ -101,7 +101,7 @@ container-build: module-check
 
 # no build lock to allow for read-only mounts
 container-serve: module-check ## Boot the development server using container.
-       $(CONTAINER_RUN) --cap-drop=ALL --cap-add=AUDIT_WRITE --read-only --mount type=tmpfs,destination=/tmp,tmpfs-mode=01777 -p 1313:1313 $(CONTAINER_IMAGE) hugo server --buildFuture --environment development --themesDir /usr/local/lib/node_modules --bind 0.0.0.0 --destination /tmp/hugo --cleanDestinationDir --noBuildLock
+       $(CONTAINER_RUN) --cap-drop=ALL --cap-add=AUDIT_WRITE --read-only --mount type=tmpfs,destination=/tmp,tmpfs-mode=01777 -p 1313:1313 $(CONTAINER_IMAGE) hugo server --buildFuture --environment development --themesDir /npm-dep/node_modules --bind 0.0.0.0 --destination /tmp/hugo --cleanDestinationDir --noBuildLock
 
 test-examples:
        scripts/test_examples.sh install
```

File: Makefile [hunk 2]
```diff
@@ -97,11 +97,11 @@ docker-push: ## Build a multi-architecture image and push that into the registry
 	rm Dockerfile.cross
 
 container-build: module-check
-	$(CONTAINER_RUN) --read-only --mount type=tmpfs,destination=/tmp,tmpfs-mode=01777 $(CONTAINER_IMAGE) sh -c "npm ci && hugo --minify --environment development"
+	$(CONTAINER_RUN) --read-only --mount type=tmpfs,destination=/tmp,tmpfs-mode=01777 $(CONTAINER_IMAGE) sh -c "npm ci && hugo --minify --environment development --themesDir /usr/local/lib/node_modules"
```
[the remark this answers is not in the document]
Comment (sftim) [reply Makefile:100]: I'm not sure of the original intent, but typically people want the build output kept. This doesn't look like it would do that.

**Comment by chalin** [comment]
Here's something along the lines of my last proposal, though it would need to be adapted to be used here:

- #48741

If/once this gets merged, then we can see how it fits in the context of the Docsy upgrade.

**Comment by chalin** [comment]
@sftim - will gladly take another look once this gets rebased to pick up merge of 0.3 and the container-build fix.

**Comment by sftim** [comment]
Hmm, I think this needs a new approach.
/close

I hope to get time to revisit it.

**Comment by chalin** [comment]
> Hmm, I think this needs a new approach. /close
> 
> I hope to get time to revisit it.

@sftim - would you like for me to give it a try (for 0.4.x), or shall I wait for your next version of this PR. Either way, if it's this week, I'll find time to help.

**Comment by sftim** [comment]
@chalin it's a [busy time](https://www.globenewswire.com/news-release/2024/11/18/2982908/0/en/Ten10-Strengthens-AWS-Cloud-Services-Leadership-with-Acquisition-of-The-Scale-Factory.html) at work - feel free to help, I'd / we'd appreciate it
You are welcome to cherry pick any of these commits, of course

**Comment by chalin** [comment]
Great! FYI, I'll be working on:

- #48807

**Pull request closed without merging**